source/OrderedTypes/PredicativeLFP.lagda

```diff
@@ -0,0 +1,1637 @@
+Ian Ray 01/09/2023.
+
+We formalize Curi's notion of abstract inductive definition (in CZF) within
+the context of a sup-lattice L with small basis B (and β : B → L). An abstract
+inductive defintion is a subset ϕ : B × L → Prop which can be thought of as a
+'inference rule' concluding b from a (in the case ϕ(b,a) holds). An inductive
+definition induces a closure condition. For example, a subset S is closed under
+ϕ if for all b : B and a : L such that (b , a) ∈ ϕ and ↓ᴮa is 'contained' in
+S then b ∈ S. Interestingly, there is an intimate connection between the
+least-closed subsets of some inductive definition ϕ and the least fixed point
+of a monotone map related to ϕ. Constructing these least-closed subsets in
+traditional foundations is somewhat grueling. Fortunately, unlike in the realm
+of set theory, inductive constructions are first class citizens in a type
+theoretic setting. Using UF + HITs we can construct the least closed subset,
+under an inductive definition ϕ, as a special quotient inductive type (QIT).
+In this file we will develop this relationship and prove a predicative version
+of the least fixed point theorem. This work follows the paper 'On Tarski's
+Fixed Point Theorem' by Giovanni Curi:
+
+Giovanni Curi. "On Tarski's fixed point theorem". In: Proc. Amer. Math. Soc
+143 (2015), pp. 4439-4455. DOI: http://doi.org/10.1090/proc/12569
+
+For a write up of the present formalization in a type theoretic setting see
+https://arxiv.org/abs/2401.00841
+
+The content of the present formalization works towards the following result.
+
+Predicative Least Fixed Point Theorem:
+Given a 𝓥-sup-lattice L with a 𝓥-presentation and a monotone
+endomap f : L → L. If there exists a bounded abstract inductive definition
+ϕ such that f ＝ Γ ϕ, then f has a least fixed point.
+
+\begin{code}
+
+{-# OPTIONS --safe --without-K --exact-split #-}
+
+open import MLTT.Spartan
+open import UF.Equiv
+open import UF.Equiv-FunExt
+open import UF.EquivalenceExamples
+open import UF.FunExt
+open import UF.Hedberg
+open import UF.Logic
+open import UF.Powerset-MultiUniverse
+open import UF.PropTrunc
+open import UF.Retracts
+open import UF.Sets
+open import UF.Subsingletons
+open import UF.Subsingletons-FunExt
+open import UF.SubtypeClassifier
+open import UF.Size
+open import UF.SmallnessProperties
+open import UF.UniverseEmbedding
+
+module OrderedTypes.PredicativeLFP
+        (pt : propositional-truncations-exist)
+        (fe : Fun-Ext)
+        (fe' : FunExt)
```

Comment:
Why two forms of function extensionality?

Reply:
I think different imports I used required both. I am aware I could get a term FunExt from Fun-Ext so maybe I should do that.

Reply:
Where should I put that at? @t

Reply:
Maybe you meant
private
 fe' : FunExt
 fe' 𝓤 𝓥 = fe {𝓤} {𝓥}

Reply:
Put it at the top, like is done here: https://www.cs.bham.ac.uk/~mhe/TypeTopology/Ordinals.OrdinalOfOrdinals.html